presto-hive/src/main/java/com/facebook/presto/hive/FilteringPageSource.java

```diff
@@ -0,0 +1,334 @@
+            RowExpression remainingPredicate,
+            TypeManager typeManager,
+            RowExpressionService rowExpressionService,
+            ConnectorSession session,
+            ConnectorPageSource delegate)
+    {
+        this.hivePageSource = delegate;
+        this.columnMappings = columnMappings;
+        this.columns = columns;
+        this.predicate = predicate;
+        this.remainingPredicate = remainingPredicate;
+        this.typeManager = typeManager;
+        this.rowExpressionService = rowExpressionService;
+        this.session = session;
+        this.functionInputs = IntStream.range(0, columns.size())
+                .boxed()
+                .collect(toImmutableMap(i -> columns.get(i).getHiveColumnIndex(), Function.identity()));
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page = hivePageSource.getNextPage();
+        if (page == null || page.getPositionCount() == 0) {
+            return page;
+        }
+
+        int[] positions = new int[page.getPositionCount()];
+        int positionCount = page.getPositionCount();
+        for (int i = 0; i < positions.length; i++) {
+            positions[i] = i;
+        }
+
+        ImmutableMap<Integer, ColumnDomain<HiveColumnHandle>> filters = ImmutableMap.of();
+        Optional<List<ColumnDomain<HiveColumnHandle>>> columnDomains = predicate.getColumnDomains();
+        if (columnDomains.isPresent()) {
+            filters = columnDomains.get().stream().collect(toImmutableMap(col -> col.getColumn().getHiveColumnIndex(), col -> col));
+        }
+
+        Block[] blocks = new Block[page.getChannelCount()];
+        for (int i = 0; i < page.getChannelCount(); i++) {
+            blocks[i] = page.getBlock(i);
+            int columnHandle = columnMappings.get(i).getHiveColumnHandle().getHiveColumnIndex();
+            if (filters.containsKey(columnHandle)) {
+                positionCount =
+                        filterBlock(
+                                blocks[i],
+                                filters.get(columnHandle).getColumn().getHiveType().getType(typeManager),
+                                toFilter(filters.get(columnHandle).getDomain()),
+                                positions,
+                                positionCount);
+            }
+        }
+
+        if (remainingPredicate != null && !(remainingPredicate.equals(TRUE_CONSTANT))) {
+            Map<VariableReferenceExpression, InputReferenceExpression> variableToInput = columns.stream()
+                    .collect(toImmutableMap(
+                            hiveColumnIndex -> new VariableReferenceExpression(hiveColumnIndex.getName(), hiveColumnIndex.getHiveType().getType(typeManager)),
+                            hiveColumnIndex -> new InputReferenceExpression(hiveColumnIndex.getHiveColumnIndex(), hiveColumnIndex.getHiveType().getType(typeManager))));
+
+            List<FilterFunction> filterFunctions =
+                    toFilterFunctions(
```

Comment:
no need to use `toFilterFunctions` to break up the filter into conjuncts over same subsets of the columns; just make one filter function:

```
RowExpression expression = replaceExpression(remainingPredicate, variableToInput);

FilterFunction filterFunction = new FilterFunction(session,
                    rowExpressionService.getDeterminismEvaluator().isDeterministic(expression),
                    rowExpressionService.getPredicateCompiler().compilePredicate(session.getSqlFunctionProperties(), expression).get());
```